### ISSUE TITLE
[aws-stepfunctions] Please add support for new Condition operators for step functions

### ISSUE THREAD

**Issue by michaelwiles** [aws/aws-cdk#9821]
Step functions now have [more "choices", json.stringify, string concatenation and global access to the context object (and a few other things).
](https://aws.amazon.com/blogs/aws/aws-step-functions-adds-updates-to-choice-state-global-access-to-context-object-dynamic-timeouts-result-selection-and-intrinsic-functions-to-amazon-states-languages)
Please can we get this included in cdk...

### Use Case

These are _very_ useful features that have been added.

### Proposed Solution

Guess we would need to discuss exactly how these features get added.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by shivlaks** [comment]
hey @michaelwiles this is on my radar as well. let's write out the specific use cases we want to unlock / enumerate them by feature so we don't end up with a giant PR that tries to do too much.

I'm also happy to work on some of these features so we can parallelize effectively and avoid duplicating effort. let me know what you think!

**Comment by michaelwiles** [comment]
Great. That does make sense: Maybe we should refocus this issue to a narrower scope then...

I can start by saying that I need/want the new comparision operator exists in my cdk code based step function - I also want to use string construction and the Json.stringify methods. So I'd like to start with those.

My proposal is to add the following to the [Condition](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-stepfunctions/lib/condition.ts) class:

```
public static isNull(input: string): Condition
public static isNotNull(input: string): Condition 
public static isString(variable: string): Condition
public static isNotString(variable: string): Condition
public static isNumeric(variable: string): Condition
public static isNotNumeric(variable: string): Condition
public static isBoolean(variable: string): Condition
public static isNotBoolean(variable: string): Condition
public static isTimestamp(variable: string): Condition
public static isNotTimestamp(variable: string): Condition
```

I did consider compressing the isNull into one method (where value would be true for Null check and false for Not Null check)

`public static isNull(input: string, value:boolean): Condition`

But I think it's a bit less intuitive and not so clear even though it's less verbose.

Then I propose to introduce a States class - analogous to the existing JsonPath class - where we can have States.format, States.jsonToString and States.stringToJson

There is going to be a problem with for example, DynamoAttributeValue which currently validates that what is passed starts with $ and if we want to support these new functions we're going to need to relax this validation somehow (or introduce another method). Maybe we can validate that it starts with '$' or 'States.' ?

**Comment by michaelwiles** [comment]
Okay so I've started implementing all the new conditions.

Going to go ahead and create condition methods for all these. There are a lot as I'm going to add a Path option for each one. For string you could potentially reuse the existing method but not for number and boolean etc as they take a number and boolean.

I suspect it would be prudent to narrow this particular issue down to only adding the conditional operators and create another ticket for the other ones. That way we can link the pull request with this ticket.

See [here](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-choice-state.html) for all the available conditional operators.

**Comment by kennu** [comment]
I have a wish to be able to add custom attributes to the generated JSON definition. We have CustomState, but it is very complicated to use it to define e.g. your own Choice structure from scratch, just to add one missing attribute. With custom attributes we could use new features like ResultSelector before they are added to AWS CDK.

**Comment by michaelwiles** [comment]
I have lodged a [pull request](https://github.com/aws/aws-cdk/pull/9920) for the new Condition operators. Not sure if it's been seen as yet because it's been a little while...

**Comment by shivlaks** [comment]
@michaelwiles - ack, I'll take a look at the PR and also loop a couple of members from the Step Functions team to give it a glance.

**Comment by colvint** [comment]
Thanks for the PR @michaelwiles 👍 FWIW, in the meantime (before it lands) this is what we're doing:

```typescript
import { Condition } from "@aws-cdk/aws-stepfunctions"

export class IsNullCondition extends Condition {
  constructor(private readonly variable: string, private readonly value: boolean) {
    super()
  }

  public renderCondition(): any {
    return {
      Variable: this.variable,
      IsNull: this.value
    }
  }
}

export class IsPresentCondition extends Condition {
  constructor(private readonly variable: string, private readonly value: boolean) {
    super()
  }

  public renderCondition(): any {
    return {
      Variable: this.variable,
      IsPresent: this.value
    }
  }
}

// and then use it like
const definition = yourChoiceState
  .when(new IsPresentCondition('$.potentiallyMissingAttr', false), someOtherState)
  .otherwise(yetAnotherState)
```

**Comment by michaelwiles** [comment]
Pull request [merged](https://github.com/aws/aws-cdk/pull/9920) and released as part of version [1.64](https://github.com/aws/aws-cdk/releases/tag/v1.64.0)

**Comment by colvint** [comment]
> Pull request [merged](https://github.com/aws/aws-cdk/pull/9920) and released as part of version [1.64](https://github.com/aws/aws-cdk/releases/tag/v1.64.0)

Nice work @michaelwiles -- appreciate you pushing this through and making it available. Removing our extensions in favor of your new conditions. Cheers!
